packages/terra-aggregator/src/Aggregator.jsx

```diff
@@ -0,0 +1,229 @@
+      return Promise.reject();
+    }
+
+    return Promise.resolve()
+      .then(() => {
+        /**
+         * If forceDismissInstance is present, a disclosure must have been opened by the currently focused
+         * Aggregator item. Therefore, we will call the forceDismissInstance in order to keep things in sync. The promise
+         * returned by forceDismissInstance will be inserted into the Promise chain. If the promise is rejected,
+         * the Aggregator's focus state will not be reset.
+         *
+         * The focus is only reset if the disclosure was dismissed successfully.
+         */
+        if (this.forceDismissInstance) {
+          return this.forceDismissInstance().then(() => {
+            this.resetFocusState();
+          });
+        }
+
+        // If a previous disclosure is not detected, we can immediately resolve and reset the focus.
+        return Promise.resolve().then(() => {
+          this.resetFocusState();
+        });
+      });
+  }
+
+  renderItems() {
+    const { items } = this.props;
+    const { focusedItemId, focusedItemState } = this.state;
+
+    return items.map((item) => {
+      const childIsActive = focusedItemId === item.key;
+
+      /**
+       * Each child given to the Aggregator is provided with an 'aggregatorDelegate' prop with the following values:
+       * hasFocus - A Boolean flag indicating whether or not the child is currently focused
+       * requestFocus - A function that will attempt to provide focus to the calling child. It takes an Object parameter that
+       *                should hold state data relevant to the focus event. The function returns a Promise that is resolved if
+       *                the focus request was successful. The Promise is resolved with a 'disclose' function that can be used to
+       *                disclose further content in a manner managed by the Aggregator. If the focus request was unsuccessful, the
+       *                Promise will be rejected.
+       * releaseFocus - A function that will attempt to release the focus held by the calling child. Returns a promse that is
+       *                resolved if the release request was successful. If the release request was unsuccessful, the
+       *                Promise will be rejected. This function is only provided to components that are focused.
+       * itemState     - An Object containing the state given to the Aggregator during the focus request.
+       */
+      return React.cloneElement(item.component, {
+        key: item.key,
+        aggregatorDelegate: {
+          hasFocus: childIsActive,
+          requestFocus: state => this.requestFocus(item.key, state),
+          releaseFocus: childIsActive ? () => (this.releaseFocus(item.key)) : undefined,
+          itemState: childIsActive ? focusedItemState : undefined,
+        },
+      });
+    });
+  }
+
+  render() {
+    const renderedItems = this.renderItems();
```

Comment:
Is there any good way to avoid unnecessary renders when there is a selection change and it doesn't affect a particular child item?

Reply:
The items themselves could implement 'shouldComponentUpdate' in order to block rerendering of their content. 

I'd like to avoid caching child instances in state, but that would be an option too if performance proves to be a concern. Until then, I say roll with it.

Reply:
![fry](https://user-images.githubusercontent.com/34603098/35698894-09a1cd9c-0754-11e8-8527-7031017a2ec0.png)

Reply:
I'd at least add that to the documentation.